### PR TITLE
fix(progress-bar): remove data url from progress-bar's css styling

### LINKED ISSUE #8876: Buffer progress bar "ball" is blocked by CSP due to data url svg
#### Bug, feature request, or proposal:
[Currently](https://github.com/angular/material2/blob/e650b049af78fb3357388bd8db9901b251bbc467/src/lib/progress-bar/_progress-bar-theme.scss#L53-L81) the buffer bar has inlined SVG's loaded with a data-url, however this is blocked when content security policy are set with the directive `"img-src 'self'"`.

`Refused to load the image 'data:image/svg+xml;charset=UTF-8,%3Csvg%20version%3D[...]' because it violates the following Content Security Policy directive: "img-src 'self'".`

#### What is the expected behavior?
All assets should be able to load with tight CSP configuration

#### What is the current behavior?
Assets loaded via data url are blocked by CSP

#### What are the steps to reproduce?
(this cannot be reproduced in Stackblitz etc due to CSP being set by webserver or within `<head>`)

1. Set the CSP to contain `img-src 'self'`
e.g.
`content-security-policy:default-src 'none';object-src 'self';manifest-src 'self';script-src 'self';img-src 'self';style-src 'self' 'unsafe-inline';frame-src 'self';worker-src 'self' blob:`
2. Insert a `<mat-progress-bar mode="buffer"></mat-progress-bar>`
3. See the inlined svg is blocked by the policy
#### What is the use-case or motivation for changing an existing behavior?
[MDN states that](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/img-src) allowing `data:` urls is insecure:
> <scheme-source>
> A schema such as 'http:' or 'https:'. The colon is required, single quotes shouldn't be used. You can also specify data schemas (not recommended).
> data: Allows data: URIs to be used as a content source. This is insecure; an attacker can also inject arbitrary data: URIs. Use this sparingly and definitely not for scripts.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Since at least Material 5.0.0

#### Is there anything else we should know?
This is already marked with a `TODO`, but it is good to have an issue to track the reason why the TODO should be actioned.

Given the content of the svg is simply a circle, it is probable that the same effect can be achieved with a rounded block element without causing any kind of breaking change.

### PR DESCRIPTION
**Pull request by josephperrott**
Fixes #8876

### CODE REVIEW

File: src/lib/progress-bar/progress-bar.html [hunk 1]
```diff
@@ -1,5 +1,17 @@
-<!-- The background div is named as such because it appears below the other divs and is not sized based on values. -->
-<div class="mat-progress-bar-background mat-progress-bar-element"></div>
+<!--
+  The background div is named as such because it appears below the other divs and is not sized based
+  on values.
+-->
+<div class="mat-progress-bar-background mat-progress-bar-element">
+  <svg width="100%" height="5">
+    <defs>
+      <pattern id="circle" x="5" y="0" width="10" height="5" patternUnits="userSpaceOnUse">
```
Comment (crisbeto) [inline src/lib/progress-bar/progress-bar.html]: AFAIK this ID has to be unique.

File: src/lib/progress-bar/progress-bar.ts [hunk 2]
```diff
@@ -79,6 +87,19 @@ export class MatProgressBar {
...
   }
+
+  /**
+   * Reappends the already attached circle SVG element in DOM to update the color styling.
+   *
+   * For colors to take effect in the background div, the svg's <circle> element, must be removed
+   * then placed back into the DOM again.  This is necessary as the browser only defines the
+   * styling of SVG pattern elements as it is placed into the DOM.
+   */
+  private _reappendCircleSvgElement() {
```
Comment (crisbeto) [inline src/lib/progress-bar/progress-bar.ts]: I'm not convinced that this is necessary. The progress spinner supports changing the colors without having to re-append the element to the DOM. Are you sure that it doesn't have something to do with the circles not having a unique id?

Comment (josephperrott) [reply src/lib/progress-bar/progress-bar.ts]: I think you are right

File: src/lib/progress-bar/progress-bar.html [hunk 3]
```diff
@@ -1,5 +1,17 @@
-<!-- The background div is named as such because it appears below the other divs and is not sized based on values. -->
-<div class="mat-progress-bar-background mat-progress-bar-element"></div>
+<!--
+  The background div is named as such because it appears below the other divs and is not sized based
+  on values.
+-->
+<div class="mat-progress-bar-background mat-progress-bar-element">
+  <svg width="100%" height="5">
+    <defs>
+      <pattern id="{{progressbarId}}" x="5" y="0" width="10" height="5" patternUnits="userSpaceOnUse">
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.html]: Use `[id]` instead of interpolation?

File: src/lib/progress-bar/progress-bar.scss [hunk 4]
```diff
@@ -26,9 +26,12 @@ $mat-progress-bar-piece-animation-duration: 250ms !default;
   // The progress bar background is used to show the bubble animation scrolling behind a
   // buffering progress bar.
   .mat-progress-bar-background {
-    background-repeat: repeat-x;
-    background-size: 10px 4px;
-    display: none;
+    svg {
+      position: absolute;
```
Comment (jelbourn) [inline src/lib/progress-bar/progress-bar.scss]: Also comment on why the element is `absolute`?

**Comment by josephperrott** [comment]
@jelbourn ready for another look when you have a moment.

**Pull request merged**